Replace and deprecate START

pontusmelke opened the pull request:
Add built-in procedures `db.nodeManualIndexSeek`, `db.nodeManualIndexSearch`, `db.relationshipManualIndexSeek`, and `db.relationshipManualIndexSearch` replacing `START` for reading from legacy index. Also deprecate `START` and add a notification that tell the user how to rewrite the query to use a procedure or `MATCH`.

changelog:  Add procedures for reading data from manual indexes. Replaces `START n=node:index(...)` and `START r =relationship:index(...)`.

pontusmelke committed:
Procedure for seeking legacy index for nodes

pontusmelke committed:
Procedure for searching legacy index for nodes

pontusmelke committed:
Procedure for seeking legacy index for relationships

pontusmelke committed:
Procedure for searching legacy index for relationships

pontusmelke committed:
Deprecate START

Deprecate START and provide a notifications detailing how to rewrite
the query.

pontusmelke committed:
Changed method to mode=READ

It is important that a read-only user can use the index for reading.

chrisvest commented:
They're called _manual indexes_ now: https://neo4j.com/docs/java-reference/3.2/#tutorials-java-embedded-index

pontusmelke committed:
Changed procedure names to use manual index

The outwards facing name is manual indexes not legacy index.

pontusmelke commented:
@chrisvest We updated the name of the procedures. ~We also renamed the exception and the status code, however I am unsure if we can do this renaming in 3.3?~ We should rename the status code and the exception as well but I am not sure what is the appropriate deprecation cycle for doing so.

systay commented on community/cypher/frontend-3.3/src/main/scala/org/neo4j/cypher/internal/frontend/v3_3/ast/Clause.scala:
```diff
@@ -86,15 +86,46 @@ case class EmitGraph(graphName: Expression)(val position: InputPosition) extends
+  private def rewrittenQuery: String = {
+    val rewritten = items.map {
+      case NodeByIdentifiedIndex(variable, index, key, expression) =>
+        s"CALL db.nodeLegacyIndexSeek('$index', '$key', '${expression.asCanonicalStringVal}') YIELD node AS ${variable.asCanonicalStringVal}"
```
This should talk about manual indexes and not legacy stuff, right? Other examples in the next couple of lines

systay commented on community/kernel/src/main/java/org/neo4j/graphdb/impl/notification/NotificationCode.java:
```diff
@@ -184,7 +184,12 @@
             SeverityLevel.WARNING,
             Status.Statement.PlannerUnavailableWarning,
-            "START is unsupported for current CYPHER version, the query has been execute by an older CYPHER version"
+            "START is not supported for current CYPHER version, the query has been executed by an older CYPHER version"
```
Is this still used or lying around for compatibility reasons? I thought this PR removed the fallback.

pontusmelke replied:
It is still around, just changed the wording. So my thinking was that it is two different notifications, one is saying: _this query could not be run on this version of cypher_ and the other one is saying: _`START` is deprecated_.  It is really two different things, or is my thinking wrong?

systay replied:
Makes sense. 👍
